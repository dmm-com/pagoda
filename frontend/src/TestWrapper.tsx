import { createTheme, ThemeProvider } from "@mui/material/styles";
import { SnackbarProvider } from "notistack";
import React, { FC } from "react";
import { MemoryRouter } from "react-router";

export const TestWrapper: FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const theme = createTheme();
  return (
    <ThemeProvider theme={theme}>
      <SnackbarProvider maxSnack={1} autoHideDuration={100}>
        <MemoryRouter>{children}</MemoryRouter>
      </SnackbarProvider>
    </ThemeProvider>
  );
};

export const TestWrapperWithoutRoutes: FC<{ children: React.ReactNode }> = ({
  children,
}) => {
  const theme = createTheme();
  return (
    <ThemeProvider theme={theme}>
      <SnackbarProvider maxSnack={1} autoHideDuration={100}>
        {children}
      </SnackbarProvider>
    </ThemeProvider>
  );
};
