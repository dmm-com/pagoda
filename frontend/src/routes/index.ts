export * from "./AppRouter";
export * from "./Routes";
