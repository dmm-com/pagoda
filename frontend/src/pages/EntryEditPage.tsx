import { zodResolver } from "@hookform/resolvers/zod";
import { Box } from "@mui/material";
import React, { FC, useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router";
import { v4 as uuidv4 } from "uuid";

import { useAsyncWithThrow } from "../hooks/useAsyncWithThrow";

import { Loading } from "components/common/Loading";
import { PageHeader } from "components/common/PageHeader";
import { SubmitButton } from "components/common/SubmitButton";
import { EntityBreadcrumbs } from "components/entity/EntityBreadcrumbs";
import { EntryBreadcrumbs } from "components/entry/EntryBreadcrumbs";
import {
  EntryForm as DefaultEntryForm,
  EntryFormProps,
} from "components/entry/EntryForm";
import { Schema, schema } from "components/entry/entryForm/EntryFormSchema";
import { useFormNotification } from "hooks/useFormNotification";
import { usePrompt } from "hooks/usePrompt";
import { useTypedParams } from "hooks/useTypedParams";
import { aironeApiClient } from "repository/AironeApiClient";
import { entityEntriesPath, entryDetailsPath } from "routes/Routes";
import {
  extractAPIException,
  isResponseError,
} from "services/AironeAPIErrorUtil";
import {
  convertAttrsFormatCtoS,
  formalizeEntryInfo,
} from "services/entry/Edit";

interface Props {
  excludeAttrs?: string[];
  EntryForm?: FC<EntryFormProps>;
  useUUID?: boolean;
}

export const EntryEditPage: FC<Props> = ({
  excludeAttrs = [],
  EntryForm = DefaultEntryForm,
  useUUID = false,
}) => {
  const { entityId, entryId } = useTypedParams<{
    entityId: number;
    entryId: number;
  }>();

  const willCreate = entryId == null;

  const navigate = useNavigate();
  const { enqueueSubmitResult } = useFormNotification("アイテム", willCreate);

  const [initialized, setInitialized] = useState(false);

  const {
    formState: { isValid, isDirty, isSubmitting, isSubmitSuccessful },
    handleSubmit,
    reset,
    setError,
    setValue,
    control,
  } = useForm<Schema>({
    resolver: zodResolver(schema),
    mode: "onBlur",
  });

  usePrompt(
    isDirty && !isSubmitSuccessful,
    "編集した内容は失われてしまいますが、このページを離れてもよろしいですか？",
  );

  const entity = useAsyncWithThrow(async () => {
    return await aironeApiClient.getEntity(entityId);
  });

  const entry = useAsyncWithThrow(async () => {
    return entryId != undefined
      ? await aironeApiClient.getEntry(entryId)
      : undefined;
  });

  useEffect(() => {
    if (willCreate) {
      if (!entity.loading && entity.value != null) {
        const entryInfo = formalizeEntryInfo(
          undefined,
          entity.value,
          excludeAttrs,
        );
        entryInfo.name = useUUID ? uuidv4() : "";
        reset(entryInfo);
        setInitialized(true);
      }
    } else {
      if (
        !entity.loading &&
        entity.value != null &&
        !entry.loading &&
        entry.value != null
      ) {
        const entryInfo = formalizeEntryInfo(
          entry.value,
          entity.value,
          excludeAttrs,
        );
        reset(entryInfo);
        setInitialized(true);
      }
    }
  }, [willCreate, entity.value, entry.value]);

  useEffect(() => {
    if (isSubmitSuccessful) {
      if (willCreate) {
        navigate(entityEntriesPath(entityId), { replace: true });
      } else {
        navigate(entryDetailsPath(entityId, entryId), { replace: true });
      }
    }
  }, [isSubmitSuccessful]);

  const handleSubmitOnValid = async (entry: Schema) => {
    const updatedAttr = convertAttrsFormatCtoS(entry.attrs);

    try {
      if (willCreate) {
        await aironeApiClient.createEntry(entityId, entry.name, updatedAttr);
      } else {
        await aironeApiClient.updateEntry(entryId, entry.name, updatedAttr);
      }
      enqueueSubmitResult(true);
    } catch (e) {
      console.log("e", e);
      if (e instanceof Error && isResponseError(e)) {
        await extractAPIException<Schema>(
          e,
          (message) => enqueueSubmitResult(false, `詳細: "${message}"`),
          (name, message) => {
            setError(name, { type: "custom", message: message });
            enqueueSubmitResult(false);
          },
        );
      } else {
        enqueueSubmitResult(false);
      }
    }
  };

  const handleCancel = () => {
    if (willCreate) {
      navigate(entityEntriesPath(entityId), { replace: true });
    } else {
      navigate(entryDetailsPath(entityId, entryId), { replace: true });
    }
  };

  if (entity.loading || entry.loading) {
    return <Loading />;
  }

  if (
    !entity.loading &&
    entity.value == undefined &&
    !entry.loading &&
    entry.value == undefined
  ) {
    throw Error("both entity and entry are invalid");
  }

  return (
    <Box>
      {entry.value ? (
        <EntryBreadcrumbs entry={entry.value} title="編集" />
      ) : (
        <EntityBreadcrumbs entity={entity.value} title="作成" />
      )}

      <PageHeader
        title={entry?.value != null ? entry.value.name : "新規アイテムの作成"}
        description={entry?.value != null ? "アイテム編集" : undefined}
      >
        <SubmitButton
          name="保存"
          disabled={!isDirty || !isValid || isSubmitting || isSubmitSuccessful}
          isSubmitting={isSubmitting}
          handleSubmit={handleSubmit(handleSubmitOnValid, (errors) => {
            console.log(errors);
          })}
          handleCancel={handleCancel}
        />
      </PageHeader>

      {initialized && entity.value != null && (
        <EntryForm
          entity={{
            ...entity.value,
            attrs: entity.value.attrs.filter(
              (attr) => !excludeAttrs.includes(attr.name),
            ),
          }}
          control={control}
          setValue={setValue}
        />
      )}
    </Box>
  );
};
