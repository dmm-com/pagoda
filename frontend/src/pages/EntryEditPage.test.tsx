/**
 * @jest-environment jsdom
 */

import { render, screen, act, waitFor } from "@testing-library/react";
import { http, HttpResponse } from "msw";
import { setupServer } from "msw/node";
import { createMemoryRouter, RouterProvider } from "react-router";

import { EntryEditPage } from "./EntryEditPage";

import { TestWrapperWithoutRoutes } from "TestWrapper";
import { entryEditPath } from "routes/Routes";

const server = setupServer(
  // getEntity
  http.get("http://localhost/entity/api/v2/2/", () => {
    return HttpResponse.json({
      id: 2,
      name: "test entity",
      note: "",
      isToplevel: false,
      attrs: [],
      webhooks: [],
    });
  }),
  // getEntry
  http.get("http://localhost/entry/api/v2/1/", () => {
    return HttpResponse.json({
      id: 1,
      name: "test entry",
      isActive: true,
      schema: {
        id: 2,
        name: "test entity",
      },
      attrs: [],
    });
  }),
);

beforeAll(() => server.listen());

afterEach(() => server.resetHandlers());

afterAll(() => server.close());

describe("EntryEditPage", () => {
  Object.defineProperty(window, "django_context", {
    value: {
      user: {
        is_superuser: false,
      },
    },
    writable: false,
  });

  test("should match snapshot", async () => {
    const router = createMemoryRouter(
      [
        {
          path: entryEditPath(":entityId", ":entryId"),
          element: <EntryEditPage />,
        },
      ],
      {
        initialEntries: ["/ui/entities/2/entries/1/edit"],
      },
    );
    const result = await act(async () => {
      return render(<RouterProvider router={router} />, {
        wrapper: TestWrapperWithoutRoutes,
      });
    });
    await waitFor(() => {
      expect(screen.queryByTestId("loading")).not.toBeInTheDocument();
    });

    expect(result).toMatchSnapshot();
  });
});
