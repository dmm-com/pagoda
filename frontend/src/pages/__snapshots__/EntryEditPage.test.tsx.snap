// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryEditPage should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-0"
      >
        <div
          class="MuiBox-root css-1baykhj"
        >
          <div
            class="MuiBox-root css-t0pdf1"
          >
            <div
              class="MuiBox-root css-jqffe8"
            >
              <nav
                aria-label="breadcrumb"
                class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
              >
                <ol
                  class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
                >
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/"
                    >
                      Top
                    </a>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/entities"
                    >
                      モデル一覧
                    </a>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <div
                      class="MuiBox-root css-162tvoi"
                    >
                      <a
                        class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                        data-discover="true"
                        href="/ui/entities/2/entries"
                      >
                        test entity
                      </a>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                        data-testid="LockIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2m-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1s3.1 1.39 3.1 3.1z"
                        />
                      </svg>
                    </div>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <div
                      class="MuiBox-root css-162tvoi"
                    >
                      <a
                        class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                        data-discover="true"
                        href="/ui/entities/2/entries/1/details"
                      >
                        test entry
                      </a>
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                        data-testid="LockIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2m-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1s3.1 1.39 3.1 3.1z"
                        />
                      </svg>
                    </div>
                  </li>
                  <li
                    aria-hidden="true"
                    class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                  >
                    /
                  </li>
                  <li
                    class="MuiBreadcrumbs-li"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
                    >
                      編集
                    </p>
                  </li>
                </ol>
              </nav>
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1cs9h4l"
        >
          <div
            class="MuiBox-root css-1519avi"
          >
            <div
              class="MuiBox-root css-9rtbay"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6 css-2toz8d-MuiTypography-root"
                id="title"
              >
                test entry
              </h6>
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
                id="description"
              >
                アイテム編集
              </h6>
              <div
                class="MuiBox-root css-hoapfw"
              >
                <div
                  class="MuiBox-root css-1kdfklj"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-8fe6cp-MuiButtonBase-root-MuiButton-root"
                    disabled=""
                    id=":r0:"
                    tabindex="-1"
                    type="button"
                  >
                    保存
                  </button>
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-152i8rg-MuiButtonBase-root-MuiButton-root"
                    id=":r1:"
                    tabindex="0"
                    type="button"
                  >
                    キャンセル
                  </button>
                </div>
              </div>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-lhgpb-MuiDivider-root"
            />
          </div>
        </div>
        <div
          class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
        >
          <div
            class="MuiBox-root css-1afd0kf"
          >
            <a
              class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault css-soe3uv-MuiButtonBase-root-MuiChip-root"
              href="#name"
              id="chip_name"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorDefault css-1umw9bq-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m7 10 5 5 5-5z"
                />
              </svg>
              <span
                class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
              >
                アイテム名
              </span>
            </a>
          </div>
          <table
            class="MuiTable-root css-1xwxv7r-MuiTable-root"
            id="table_attr_list"
          >
            <thead
              class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
            >
              <tr
                class="MuiTableRow-root MuiTableRow-head css-1wtsgsb-MuiTableRow-root"
              >
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-7i4wkh-MuiTableCell-root"
                  scope="col"
                >
                  項目
                </th>
                <th
                  class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-7i4wkh-MuiTableCell-root"
                  scope="col"
                >
                  内容
                </th>
              </tr>
            </thead>
            <tbody
              class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
            >
              <tr
                class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
              >
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1egp9pz-MuiTableCell-root"
                >
                  <div
                    class="MuiBox-root css-yeouz0"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-zaqce3-MuiTypography-root"
                      id="name"
                    >
                      アイテム名
                    </p>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1yompuc-MuiTypography-root"
                    >
                      必須
                    </p>
                  </div>
                </td>
                <td
                  class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1egp9pz-MuiTableCell-root"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
                  >
                    <div
                      class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-6h9cc6-MuiInputBase-root-MuiInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
                        data-1p-ignore="true"
                        id="entry-name"
                        name="name"
                        type="text"
                        value="test entry"
                      />
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
          <div
            class="MuiBox-root css-1h9v3gn"
          >
            <button
              class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-6vxf02-MuiButtonBase-root-MuiFab-root"
              id="scroll_button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="KeyboardArrowUpRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.12 14.71 12 10.83l3.88 3.88c.39.39 1.02.39 1.41 0s.39-1.02 0-1.41L12.7 8.71a.996.996 0 0 0-1.41 0L6.7 13.3c-.39.39-.39 1.02 0 1.41.39.38 1.03.39 1.42 0"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-0"
    >
      <div
        class="MuiBox-root css-1baykhj"
      >
        <div
          class="MuiBox-root css-t0pdf1"
        >
          <div
            class="MuiBox-root css-jqffe8"
          >
            <nav
              aria-label="breadcrumb"
              class="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-5cvx94-MuiTypography-root-MuiBreadcrumbs-root"
            >
              <ol
                class="MuiBreadcrumbs-ol css-1uwp4ue-MuiBreadcrumbs-ol"
              >
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                    data-discover="true"
                    href="/ui/"
                  >
                    Top
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                    data-discover="true"
                    href="/ui/entities"
                  >
                    モデル一覧
                  </a>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <div
                    class="MuiBox-root css-162tvoi"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/entities/2/entries"
                    >
                      test entity
                    </a>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="LockIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2m-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1s3.1 1.39 3.1 3.1z"
                      />
                    </svg>
                  </div>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <div
                    class="MuiBox-root css-162tvoi"
                  >
                    <a
                      class="MuiTypography-root MuiTypography-body1 css-1rk5fdy-MuiTypography-root"
                      data-discover="true"
                      href="/ui/entities/2/entries/1/details"
                    >
                      test entry
                    </a>
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                      data-testid="LockIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2m-6 9c-1.1 0-2-.9-2-2s.9-2 2-2 2 .9 2 2-.9 2-2 2m3.1-9H8.9V6c0-1.71 1.39-3.1 3.1-3.1s3.1 1.39 3.1 3.1z"
                      />
                    </svg>
                  </div>
                </li>
                <li
                  aria-hidden="true"
                  class="MuiBreadcrumbs-separator css-jhr02i-MuiBreadcrumbs-separator"
                >
                  /
                </li>
                <li
                  class="MuiBreadcrumbs-li"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1bgbmvw-MuiTypography-root"
                  >
                    編集
                  </p>
                </li>
              </ol>
            </nav>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-1cs9h4l"
      >
        <div
          class="MuiBox-root css-1519avi"
        >
          <div
            class="MuiBox-root css-9rtbay"
          >
            <h6
              class="MuiTypography-root MuiTypography-h6 css-2toz8d-MuiTypography-root"
              id="title"
            >
              test entry
            </h6>
            <h6
              class="MuiTypography-root MuiTypography-subtitle1 css-1i3tsf3-MuiTypography-root"
              id="description"
            >
              アイテム編集
            </h6>
            <div
              class="MuiBox-root css-hoapfw"
            >
              <div
                class="MuiBox-root css-1kdfklj"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-8fe6cp-MuiButtonBase-root-MuiButton-root"
                  disabled=""
                  id=":r0:"
                  tabindex="-1"
                  type="button"
                >
                  保存
                </button>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo MuiButton-root MuiButton-contained MuiButton-containedInfo MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorInfo css-152i8rg-MuiButtonBase-root-MuiButton-root"
                  id=":r1:"
                  tabindex="0"
                  type="button"
                >
                  キャンセル
                </button>
              </div>
            </div>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-lhgpb-MuiDivider-root"
          />
        </div>
      </div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-5c1adp-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-1afd0kf"
        >
          <a
            class="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-sizeMedium MuiChip-colorDefault MuiChip-clickable MuiChip-clickableColorDefault MuiChip-outlinedDefault css-soe3uv-MuiButtonBase-root-MuiChip-root"
            href="#name"
            id="chip_name"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiChip-icon MuiChip-iconMedium MuiChip-iconColorDefault css-1umw9bq-MuiSvgIcon-root"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m7 10 5 5 5-5z"
              />
            </svg>
            <span
              class="MuiChip-label MuiChip-labelMedium css-16cgrcw-MuiChip-label"
            >
              アイテム名
            </span>
          </a>
        </div>
        <table
          class="MuiTable-root css-1xwxv7r-MuiTable-root"
          id="table_attr_list"
        >
          <thead
            class="MuiTableHead-root css-1a7iywq-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1wtsgsb-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-7i4wkh-MuiTableCell-root"
                scope="col"
              >
                項目
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-7i4wkh-MuiTableCell-root"
                scope="col"
              >
                内容
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-gmh7jj-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-1xtozoh-MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1egp9pz-MuiTableCell-root"
              >
                <div
                  class="MuiBox-root css-yeouz0"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-zaqce3-MuiTypography-root"
                    id="name"
                  >
                    アイテム名
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1yompuc-MuiTypography-root"
                  >
                    必須
                  </p>
                </div>
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1egp9pz-MuiTableCell-root"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
                >
                  <div
                    class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-6h9cc6-MuiInputBase-root-MuiInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiInput-input css-1yrc8ca-MuiInputBase-input-MuiInput-input"
                      data-1p-ignore="true"
                      id="entry-name"
                      name="name"
                      type="text"
                      value="test entry"
                    />
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="MuiBox-root css-1h9v3gn"
        >
          <button
            class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-6vxf02-MuiButtonBase-root-MuiFab-root"
            id="scroll_button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
              data-testid="KeyboardArrowUpRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.12 14.71 12 10.83l3.88 3.88c.39.39 1.02.39 1.41 0s.39-1.02 0-1.41L12.7 8.71a.996.996 0 0 0-1.41 0L6.7 13.3c-.39.39-.39 1.02 0 1.41.39.38 1.03.39 1.42 0"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
